Game Interface ported to T5.

xerxes-at opened the pull request:
Splits the game interface into generic and game specific code.
Ports the game interface with virtually all functionality to T5MP.
Adds a deploy bat to quickly copy all gsc files to the default folders of Pluto.
Adds basic logging logic which can be extended per game depending on its capabilities (eg. logging to the server console)

xerxes-at committed:
Split up integration logic based on game and add support for PlutoT5.

* Splits _integration.gsc into _integration.gsc and _integration_iw4_iw5.gsc
* Adds _integration_t5.gsc
* Adds deploy.bat to deploy the files to the default Pluto folder.

xerxes-at committed:
Meme fix to get T5 working again.

RaidMax commented on GameFiles/_integration.gsc:
```diff
@@ -38,13 +34,35 @@ init()
     level.clientCommandRusAsTarget = [];
+    level.overrideMethods = [];
+
+    wait 0.1;
```
What's the purpose of this wait? It's not immediately clear so I would add a comment of why we're doing this.

xerxes-at replied:
That is just some leftover, removed with 7510ebd

RaidMax commented on GameFiles/_integration.gsc:
```diff
@@ -38,13 +34,35 @@ init()
+    InitializeLogger();
+
+    //The gamename dvar seems to be an IW only thing..
+    if (level.eventBus.gamename == "") {
```
format: space between ( ) and condition

RaidMax commented on GameFiles/_integration.gsc:
```diff
@@ -38,13 +34,35 @@ init()
+    [[level.overrideMethods["SetDvarIfUninitialized"]]]( "sv_iw4madmin_integration_enabled", 1 );
+    [[level.overrideMethods["SetDvarIfUninitialized"]]]( "sv_iw4madmin_integration_debug", 0 );
+
+    level.logger = spawnstruct();
```
I would move this up with the other items initialized above line 37 etc...

RaidMax commented on GameFiles/_integration.gsc:
```diff
@@ -38,13 +34,35 @@ init()
+    //The gamename dvar seems to be an IW only thing..
+    if (level.eventBus.gamename == "") {
+        version = getDvar( "version" );
+        switch( version )
+	    {
+            case "Call of Duty Multiplayer - Ship COD_T5_S MP build 7.0.189 CL(1022875) CODPCAB-V64 CEG Wed Nov 02 18:02:23 2011 win-x86":
+                level.eventBus.gamename = "T5MP";
+            break;
+        }
```
instead of doing switch here let's create a dictionary with the version pointing to the game name as this could in theory could get quite large
eg:
```
level.gameNames = [];
level.gameNames["Call of Duty Multiplayer - Ship COD_T5_S MP build 7.0.189 CL(1022875) CODPCAB-V64 CEG Wed Nov 02 18:02:23 2011 win-x86"] = "T5MP";
```

The default will be an empty string if the dict key doesn't exist so we can catch any logic there

RaidMax replied:
format: spacing/indention is inconsistent here. Should be 4 spaces for every indent level, no tabs.
VS code should able to convert from tabs to spaces to make it easier if you're using tabs anywhere

RaidMax commented on GameFiles/_integration.gsc:
```diff
@@ -38,13 +34,35 @@ init()
+        }
+    }
     
     if ( GetDvarInt( "sv_iw4madmin_integration_enabled" ) != 1 )
```
might as well move this up as far as possible since we don't need to do any additional logic if it's disabled

RaidMax commented on GameFiles/_integration.gsc:
```diff
@@ -165,8 +184,8 @@ DisplayWelcomeData()
-    guidHigh = self GetPlayerData( "bests", "none" ); 
-    guidLow = self GetPlayerData( "awards", "none" );
+    guidHigh = self [[level.overrideMethods["GetPlayerData"]]]( "bests", "none" ); 
+    guidLow =  self [[level.overrideMethods["GetPlayerData"]]]( "awards", "none" );
```
nit: remove extra space

RaidMax commented on GameFiles/_integration.gsc:
```diff
@@ -278,27 +297,71 @@ RegisterClientCommands()
-    
-    if ( isDefined( ::God ) )
+    level.overrideMethods["UnsupportedFunc"]= ::UnsupportedFunc;
+    level notify("InitializeGameMethods");
+    level waittill("InitializeGameMethodsDone");
+}
```
format ( space )

RaidMax commented on GameFiles/_integration.gsc:
```diff
@@ -278,27 +297,71 @@ RegisterClientCommands()
-        level.overrideMethods["noclip"] = ::NoClip;
+}
+
+LogWarning( message )
```
might want to add a log debug here so we can reuse this method every place I do a check against whether debug is enabled or not.
eg: 
![image](https://user-images.githubusercontent.com/6732771/179421656-c9734439-9ddc-41bd-bb23-83f2c95940ba.png)

becomes `LogDebug( "...." )`

RaidMax commented on GameFiles/_integration.gsc:
```diff
@@ -278,27 +297,71 @@ RegisterClientCommands()
+
+Log2Dvar( LogLevel, message )
+{
+    switch (LogLevel) {
```
format: ( LogLevel )
inconsistent brace new line

diamante0018 replied:
I would like you to use `createprintchannel` builtin to create different channels for the various levels of logging.
Do not use IPrintLn. Use `print` or `printf`

xerxes-at replied:
Will that work on all cod titles? _integration.gsc should be as generic as possible so it can be used in as many cod titles as possible without changes. New loggers that use game specific functionality can always be registered in the game specific files.

diamante0018 replied:
print will work on T5, T6 andIW5.
On T4/IW4 it will only work with developer_script set to 1. Alternative: printconsole.
createprintchannel and setprintchannel are as generic as it gets.
See utility.gsc

RaidMax commented on GameFiles/_integration.gsc:
```diff
@@ -278,27 +297,71 @@ RegisterClientCommands()
-        level.overrideMethods["noclip"] = ::IW5_NoClip;
+Log2IngamePrint( LogLevel, message )
+{
+    switch (LogLevel) {
+    case "error":
```
format: ( LogLevel )
inconsistent brace new line
case indention

RaidMax commented on GameFiles/_integration.gsc:
```diff
@@ -485,17 +548,20 @@ SetClientMeta( metaKey, metaValue, clientId, direction )
 
 SaveTrackingMetrics()
 {
+    if (!isDefined(self.persistentClientId)) {
```
format: 
```cpp
if ( !IsDefined( self.persistentClientId ) )
{
```

RaidMax commented on GameFiles/_integration.gsc:
```diff
@@ -485,17 +548,20 @@ SetClientMeta( metaKey, metaValue, clientId, direction )
 {
+    if (!isDefined(self.persistentClientId)) {
+        return;
+    }
```
format: add new line here

RaidMax commented on GameFiles/_integration.gsc:
```diff
@@ -587,7 +653,7 @@ QueueEvent( request, eventType, notifyEntity )
     while ( GetDvar( level.eventBus.inVar ) != "" && ( GetTime() - start ) < maxWait )
     {
-        level waittill_notify_or_timeout( "bus_ready", 1 );
+        level [[level.overrideMethods["waittill_notify_or_timeout"]]]( "bus_ready", 1 );
```
Shame this method isn't ubiquitous

RaidMax commented on GameFiles/_integration.gsc:
```diff
@@ -633,14 +699,22 @@ ParseDataString( data )
     
     counter = 0;
-    foreach ( part in dataParts )
+    if( isdefined( dataParts ) )
```
format:
```
if ( IsDefined( ....
```

RaidMax commented on GameFiles/_integration.gsc:
```diff
@@ -633,14 +699,22 @@ ParseDataString( data )
-        dict[key] = value;
-        dict[counter] = key;
-        counter++;
+        //foreach ( part in dataParts )
+        for( i = 0; i < dataParts.size; i++ )
```
How come foreach doesn't work here? can we remove the commented code if it's not usable

xerxes-at replied:
T5 doesn't know the foreach keyword.

RaidMax commented on GameFiles/_integration.gsc:
```diff
@@ -633,14 +699,22 @@ ParseDataString( data )
+        //foreach ( part in dataParts )
+        for( i = 0; i < dataParts.size; i++ )
+        {
+            if( isdefined( dataParts[i] ) )
```
format:
```
if ( IsDefined( ....
```

RaidMax commented on GameFiles/_integration.gsc:
```diff
@@ -907,8 +989,8 @@ NoClipImpl()
 
-    self call [[level.overrideMethods["god"]]]( true );
-    self call [[level.overrideMethods["noclip"]]]( true );
+    self [[level.overrideMethods["god"]]]( true );
+    self [[level.overrideMethods["noclip"]]]( true );
```
I don't think we can just remove the `call` keyword here. This was added because on IW4 it won't execute our "function pointers" without the call keyword.

diamante0018 replied:
Do not remove the word call on IW engine. Calls to builtin will not work

RaidMax commented on GameFiles/_integration.gsc:
```diff
@@ -1005,6 +1087,9 @@ AlertImpl( event, data )
     if ( level.eventBus.gamename == "IW5" ) { //IW5's notification are a bit different...
         self thread maps\mp\gametypes\_hud_message::oldNotifyMessage( data["alertType"], data["message"], undefined, ( 1, 0, 0 ), "ui_mp_nukebomb_timer", 7.5 );
     }
+    if ( level.eventBus.gamename == "T5MP" ) {
```
format: put brace on new line (I didn't catch it for the one above too)

RaidMax commented on GameFiles/_integration.gsc:
```diff
@@ -1090,8 +1175,10 @@ NightModeImpl()
     IPrintLnBold( message );
 
-    foreach( player in level.players )
+    //foreach( player in level.players )
```
remove commented code if not using

RaidMax commented on GameFiles/deploy.bat:
```diff
@@ -0,0 +1,14 @@
+
+ECHO "Pluto T6"
+xcopy /y .\PT6\storage\t6\scripts\mp\_customcallbacks.gsc "%LOCALAPPDATA%\Plutonium\storage\t6\scripts\mp"
+xcopy /y .\PT6\storage\t6\scripts\mp\_customcallbacks.gsc.src "%LOCALAPPDATA%\Plutonium\storage\t6\scripts\mp"
```
add new line at end of file

RaidMax commented on GameFiles/_integration_iw4_iw5.gsc:
```diff
@@ -0,0 +1,88 @@
+init()
+{
+
```
nit: remove extra space

RaidMax commented on GameFiles/_integration_iw4_iw5.gsc:
```diff
@@ -0,0 +1,88 @@
+    level.overrideMethods["SetDvarIfUninitialized"] = ::SetDvarIfUninitialized;
+    level.overrideMethods["GetPlayerData"] = ::GetPlayerData;
+    level.overrideMethods["SetPlayerData"] = ::SetPlayerData;
+
```
nit: remove extra new line

RaidMax commented on GameFiles/_integration.gsc:
```diff
@@ -1124,49 +1211,4 @@ SetSpectatorImpl()
-    self NoClip();
-    SetDvar( "sv_cheats", 0 );
-}
+}
```
nit: add new line at end of file

diamante0018 commented on GameFiles/_integration.gsc:
```diff
@@ -38,13 +34,35 @@ init()
+
+    //The gamename dvar seems to be an IW only thing..
+    if (level.eventBus.gamename == "") {
+        version = getDvar( "version" );
```
There is a better way to do this.
`protocol` is an integer dvar and does not require you to make string comparison and is definitely better than this.
I have attached an image containing the default value on
![Capture](https://user-images.githubusercontent.com/37080671/180640219-44e2b451-bdf7-46da-a19c-035503a1a19c.PNG)
 T5. Check that plutonium is not overriding it.

xerxes-at replied:
After discussing this with @RaidMax each game will get its own gsc file and will set the variable from there.

diamante0018 commented on GameFiles/_integration.gsc:
```diff
@@ -38,13 +34,35 @@ init()
     level.clientCommandRusAsTarget = [];
+    level.overrideMethods = [];
+
+    wait 0.1;
```
why

xerxes-at committed:
First iteration of changes.

Ayymoss commented:
Just a side note, please update the project file with the file location and name changes so it appears correctly in our IDEs. :)

https://github.com/RaidMax/IW4M-Admin/blob/release/pre/IW4MAdmin.sln#L16

xerxes-at committed:
Re-base and move IW5 + T5 files to the correct folder.

xerxes-at committed:
Update _integration_iw5.gsc to use the new interface.

Basically a 1:1 copy of the iw4x file as we already spent some time on getting both games running from the same file.

xerxes-at committed:
Update _integration_t5.gsc to use the new interface.

also add a new way of achieving god mode

xerxes-at committed:
Enable all commands but NoClip on T5.

xerxes-at committed:
Fix deploy.bat

also only load the Game Interface + AC on MP until someone with ZM Servers can test it.

xerxes-at committed:
Implement latest changes for IW5 and T5

xerxes-at committed:
Fixes as many script issues as I was able to find on IW5 and T5.
